### PR TITLE
[BD-21] feat: expose toggle state report API

### PR DESCRIPTION
**Pull request by regisb**
**Description:**

This report, previously in edx-platform, could not be exposed in other IDAs.
Here, we expose the report with a simple API. The modular ToggleStateReport
class can be overridden such that custom toggle values (such as those coming
from edx-platform's WaffleFlagCourseOverrideModel) can be added to the report.

**JIRA:** https://openedx.atlassian.net/wiki/spaces/COMM/pages/1596358943/BD-21+Toggles+Settings+Documentation

**Dependencies:** a corresponding PR will be opened on the edx-platform repo: https://github.com/edx/edx-platform/pull/26325

**Reviewers:**
- [x] @robrap 

**Merge checklist:**
- [x] All reviewers approved
- [x] CI build is green
- [x] Version bumped
- [x] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [x] Commits are squashed

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is 
      finished.
- [ ] Delete working branch (if not needed anymore)

### CODE REVIEW

**Comment by regisb** [comment]
@robrap this is ready for review. A companion edx-platform PR is coming up.

File: edx_toggles/tests/test_state.py
Comment (robrap) [inline edx_toggles/tests/test_state.py:127]: Not part of this PR, but I thought module name with None was no longer allowed? I guess we don’t have an assertion?

File: edx_toggles/toggles/state.py
Comment (robrap) [inline edx_toggles/toggles/state.py:51]: Curious of the benefit of all the single line functions, rather than just inlining the function details?

**Review by robrap — APPROVED** [review]
Thank you.
* I don’t see any blockers. 
* I wasn’t super detailed in my review, because I imagine a lot of this code didn’t change much.
* I imagine there might be some existing docs in this repo that might need to be updated based on these changes?

File: edx_toggles/tests/test_state.py
Comment (regisb) [reply edx_toggles/tests/test_state.py:127]: The `module_name` argument has to be defined, but there is no value checking. My personal opinion is that we should not be adding assert statements.

File: edx_toggles/toggles/state.py
Comment (regisb) [reply edx_toggles/toggles/state.py:51]: It makes it clear that the implementation does not depend on `self`. Granted, this is a not a huge benefit. But I like the fact that the base ToggleStateReport class is kept relatively small in terms of lines of code.

File: edx_toggles/tests/test_state.py
Comment (robrap) [reply edx_toggles/tests/test_state.py:127]: I'll explain why I would add this, and then you can explain why you wouldn't.  I want `__name__` to be used everywhere, and allowing `None` enables people to not do this, and leaves the potential of a contagion of copy/paste with a bad example. The assertion could clearly note how to resolve. I know you didn't want this to be a required field to begin with, and I feel like this may just be a continuation of the same original discussion. I also may be missing something, so let me know if that is the case.  Thank you!

File: edx_toggles/toggles/state.py
Comment (robrap) [reply edx_toggles/toggles/state.py:51]: Thanks. Just curious.

File: edx_toggles/tests/test_state.py
Comment (regisb) [reply edx_toggles/tests/test_state.py:127]: Python allows developers to pass arbitrary values as function arguments. Ensuring argument correctness is difficult, and I believe it should be the developers' responsibility to make sure that the arguments that they use are correct. In this particular case, how hard should we try to ensure correctness?

- `module_name` is "truish" (`if not module_name: ...`)
- `module_name` is a string (`if not isinstance(module_name, str): ...`)
- `module_name` is an importable module (`try importlib.import_module(module_name): ... except ModuleNotFoundError: ...`)

Let's face it: there is no way to prevent developers from bad copy-pasting. If we write `WaffleFlag(..., "testmodule1")` in a test case, then other users might copy-paste it.

Comment (robrap) [reply edx_toggles/tests/test_state.py:127]: I was thinking truish with a message of what we expect someone to use. I understand we can’t prevent every issue, but I would have prevented this. I can check what other think. I might be too much of a control freak. 🙂

I hope it is clear that this is not a blocker to this PR which has been approved. Are you waiting on me for anything for merging?

**Review by robrap — COMMENTED** [review]
@regisb: I just noticed something in the edx-platform imports from here. Did you intentionally not use the internal directory and import in the edx-toggles api, or was this just an oversight? If intentional, can you explain? Thanks.

File: edx_toggles/toggles/state.py
Comment (robrap) [inline edx_toggles/toggles/state.py:21]: Sorry if I missed it, but where are docs for recommended view code? I think we have a how to for adding toggle reporting for other IDAs, and maybe something belongs there? Not a blocker, but better to not leave all doc updates till later.

**Review by robrap — COMMENTED** [review]
More thoughts...

File: edx_toggles/tests/test_state.py
Comment (regisb) [reply edx_toggles/tests/test_state.py:127]: > I hope it is clear that this is not a blocker to this PR which has been approved. Are you waiting on me for anything for merging?

You made it clear but I did not realize it :) I'll merge right away. I won't add the module_name truish check to this PR, as this would be a backward-incompatible change.

File: edx_toggles/toggles/state.py
Comment (regisb) [reply edx_toggles/toggles/state.py:21]: It's difficult for me to recommend the best view implementation, because I'm not sure how permissions should be handled. If it were up to me, I would not resort to class-based views and would simply write:

```
def get_toggle_state_report(request):
    return JsonResponse(ToggleStateReport().as_dict())
```

**Comment by regisb** [comment]
>  I just noticed something in the edx-platform imports from here. Did you intentionally not use the internal directory and import in the edx-toggles api, or was this just an oversight? If intentional, can you explain? Thanks.

https://github.com/edx/edx-platform/pull/26325#discussion_r580704944

Do you mean we should write:

```
from edx_toggles.toggles import ToggleStateReport, get_or_create_toggle_response 
```

instead of:

```
from edx_toggles.toggles.state import ToggleStateReport, get_or_create_toggle_response 
```

?

I like the second import better, as it exposes stuff that is not core to the library, but if you prefer I can add import statements to toggles/__init__.py.

**Comment by robrap** [comment]
> Do you mean we should write:
> 
> ```
> from edx_toggles.toggles import ToggleStateReport, get_or_create_toggle_response 
> ```
> 
> instead of:
> 
> ```
> from edx_toggles.toggles.state import ToggleStateReport, get_or_create_toggle_response 
> ```
> 
> ?

I'm fine with keeping `from edx_toggles.toggles.state` if it is intentionally.  

How would you feel about moving the implementation to `https://github.com/edx/edx-toggles/tree/master/edx_toggles/toggles/internal` and adding imports into a `/state/__init__.py`? The downside of this request is obvious. The upside is that it keeps the pattern of using "internal" more obvious without relying on comments that may or may not be read.  The alternative is to try to explain all this in `state.py` and hope that people only follow this pattern when they intentionally don't want to use the top-level public api.

File: edx_toggles/toggles/state.py
Comment (robrap) [reply edx_toggles/toggles/state.py:21]: It might be fine to start by pointing to the edx-platform code simply as an example.

If you want to use a function rather than a class, you could update that in edx-platform.  See https://github.com/edx/edx-platform/pull/26325#discussion_r581341136 for details.

**Comment by robrap** [comment]
@regisb: Not sure how you feel about my last comment? Whether we land as-is, or comment the intentional decision in a docstring, or adjust like my proposal, let's find a way to land this sooner rather than later.  Thanks.

**Comment by regisb** [comment]
> How would you feel about moving the implementation to https://github.com/edx/edx-toggles/tree/master/edx_toggles/toggles/internal and adding imports into a /state/__init__.py?

If we decide to move code, I think I would prefer the following:

* All actual code in `edx_toggles/toggles/state/internal/`
* Exposed API in `edx_toggles/toggles/state/__init__.py`

I'll update this PR accordingly. If you prefer a different file hierarchy you always have the option of moving the code around later, while keeping the `edx_toggles/toggles/state/__init__.py` module.

**Comment by regisb** [comment]
@robrap plkease check this updated (and rebased) PR.

File: edx_toggles/tests/test_state.py
Comment (robrap) [reply edx_toggles/tests/test_state.py:127]: Note: this line could still be updated to use `__name__` for all the reasons I wanted to make this required. I'm not going to block on this of course, but if you are willing, you can update. Either way, you can respond and mark resolved again.  Thanks.

**Review by robrap — APPROVED** [review]

**Comment by robrap** [comment]
@regisb: Of course, approval assumes tests are fixed. :)

File: edx_toggles/tests/test_state.py
Comment (regisb) [reply edx_toggles/tests/test_state.py]: Right! I updated the argument.

**Comment by regisb** [comment]
Merging!

**Pull request merged**
